Added Printing Press recipe classes and modified config for Devoted

Aleksey-Terzi opened the pull request:
Devoted config modified in the following way:

- New factory "Printing Press"
- Recipe "Bind Written Book" moved to "Printing Press"
- New recipe "Craft Printing Plate" - new item "Printing Plate" used to print books and notes.
- New recipe "Print Book" - full copy of book (including generation) from "Printing Plate"
- New recipe "Print Pamphlet" - printing text from the first page of book to the paper.
- New recipe "Print Secure Note" - same as Pamphlet but also printing serial number of "Printing Plate"

ProgrammerDan commented:
Looks pretty good, and straightforward.

No complex "queue" simulation so it might be a bit more accessible then the prior. I think I was the only person who liked the queue :D . 

Any testing been done to date?

Maxopoly commented on src/main/java/com/github/igotyou/FactoryMod/ConfigParser.java:
```diff
@@ -810,6 +814,25 @@ private IRecipe parseRecipe(ConfigurationSection config) {
 			result = new DummyParsingRecipe(identifier, name, productionTime, null);
 			break;
+		case "PRINTINGPLATE":
+			ItemMap printingPlateInput = parseItemMap(config.getConfigurationSection("input"));
```
You dont need to parse the input again here (same for the other recipes), because pretty much all the recipes use some input it is already parsed out before this switch and you can use it with the "input" variable.


Additionally it would be good if those new recipes supported config inheritation the way the other recipes do to keep things consistent.

Aleksey-Terzi replied:
Ok, thanks for suggestion.
I've used more earlier version of FactoryMod when started development where all of this seems was absent.

Aleksey-Terzi commented:
> No complex "queue" simulation so it might be a bit more accessible then the prior. I think I was the only person who liked the queue :D .

I don't know how it worked previously so implemented it as I see it.
But if queue is a good thing - I could add it, just describe how it is working :)

> Any testing been done to date?

Yes, on local environment.

ProgrammerDan commented:
> But if queue is a good thing

Nah, 95% of people didn't understand it, and although it was cool because it emulated an actual printing press's delayed output and aggregation of input functionality, it wasn't a good fit for the average player, in terms of usability. 

What you have is great :).

Aleksey-Terzi commented:
> Nah, 95% of people didn't understand it, and although it was cool because it emulated an actual printing press's delayed output and aggregation of input functionality, it wasn't a good fit for the average player, in terms of usability.
>
> What you have is great :).

Ok, I see, thanks :)

Suggested changes are done.

Also:

- Refactored printing press recipe classes

- TagManager now used to save NBT tag data

- Fixed method FurnCraftChestFactory.getRecipeLevel. Now it checks result of **recipeLevel.get(r)** for NULL result and return 1 if it is NULL. This is done because of sometimes this code crashed on server shutdown.

ProgrammerDan commented:
This looks really good. Going to go ahead and merge, we'll do some testing later on after I get bastions out.

Maxopoly commented:
btw that one fix for recipe level get during saving is critical, it'll prevent save files from being created if you remove a recipe from a factory while keeping the recipe in the config.

ProgrammerDan commented:
Oh wow ok.
